File: packages/engine/lib/execution/src/runner/javascript/batch.js
// noinspection BadExpressionStatementJS
import { arrow } from "./lib/execution/src/runner/javascript/apache-arrow-bundle.js";
import * as hash_util from "./lib/execution/src/runner/javascript/hash_util.js";

const Batch = function () {
  this.mem_version = -1;
  this.batch_version = -1;
  this.mem = null; // After loading, `mem` will be an ArrayBuffer.
  this.vectors = {};
  this.cols = {}; // Syncing erases columns that have become invalid.
};

const get_u64 = (dataview, offset) => {
  const left = dataview.getUint32(offset, true);
  const right = dataview.getUint32(offset + 4, true);
  const combined = left + 2 ** 32 * right; // Assumes little-endian.

  // `MAX_SAFE_INTEGER` is (2^53 - 1). We (currently) don't support more
  // than 2^32 agents, so as long as we use less than about 2^20 bytes
  // per agent (including message pools, whose size can grow quadratically
  // with the number of agents in the worst case), the shared memory length
  // will accurately fit in a double.
  // Also, 2^53 bytes is 2^43 kilobytes, i.e. 2^33 megabytes, i.e. 2^23
  // gigabytes, i.e. 2^13 terabytes. It's unlikely that any HASH users
  // will have that much RAM or even disk space. Even virtual memory
  // address space is usually less than 2^53 bytes (current 64-bit
  // computers don't enable using the full 64 bits of address space).

  if (!Number.isSafeInteger(combined)) {
    throw new RangeError(combined + " exceeds MAX_SAFE_INTEGER.");
  }
  return combined;
};
const parse_any_type_fields = (metadata) => {
  const any_type_fields = new Set();

  let field_names = metadata.get("any_type_fields");

  if (field_names) {
    for (let field_name of field_names.split(",")) {
      any_type_fields.add(field_name);
    }
  }

  return any_type_fields;
};

const load_vectors = (record_batch_bytes, schema) => {
  const reader = new arrow.MessageReader(record_batch_bytes);
  const msg = reader.readMessage();
  const header = msg.header();
  const body = reader.readMessageBody(msg.bodyLength);
  const dicts = new Map();
  const loader = new reader.VectorLoader(
    body,
    header.nodes,
    header.buffers,
    dicts,
  );
  const vector_list = loader.visitMany(schema.fields);
  const any_type_fields = parse_any_type_fields(schema.metadata);
  // Unnecessary:
  // const record_batch = new arrow.RecordBatch(schema, header.length, vector_list);

  const vectors = {}; // Field name --> vector.
  for (var i = 0; i < vector_list.length; ++i) {
    // `VectorLoader` doesn't actually return instances of `Vector` for some reason.
    const vector = new arrow.makeVector(vector_list[i]);
    const field = schema.fields[i];
    vector.type.is_any = any_type_fields.has(field.name);
    vectors[field.name] = vector;
  }
  return vectors;
};

const load_markers = (shared_bytes) => {
  // `shared_bytes` should be ArrayBuffer.
  const dataview = new DataView(shared_bytes);
  const m = {
    // TODO: Use Uint32Array instead of Dataview, both
    //       here and in `get_u64`.
    // Size of `u64` is 8 bytes.
    schema_offset: get_u64(dataview, 0 * 8),
    schema_size: get_u64(dataview, 1 * 8),
    header_offset: get_u64(dataview, 2 * 8),
    header_size: get_u64(dataview, 3 * 8),
    meta_offset: get_u64(dataview, 4 * 8),
    meta_size: get_u64(dataview, 5 * 8),
    data_offset: get_u64(dataview, 6 * 8),
    data_size: get_u64(dataview, 7 * 8),
  };

  if (m.schema_offset + m.schema_size > m.header_offset) {
    throw new RangeError("schema marker");
  }
  if (m.header_offset + m.header_size > m.meta_offset) {
    throw new RangeError("header marker");
  }
  if (m.meta_offset + m.meta_size > m.data_offset) {
    throw new RangeError("meta marker");
  }
  if (m.data_offset + m.data_size > shared_bytes.length) {
    throw new RangeError("data marker");
  }
  return m;
};

const load_marked_vectors = (shared_bytes, schema) => {
  const markers = load_markers(shared_bytes); // Record batch bytes are subset of all shared.
  const data_end = markers.data_offset + markers.data_size;
  const record_batch_offset = markers.meta_offset;
  const record_batch_size = data_end - record_batch_offset;
  const record_batch_bytes = new Uint8Array(
    shared_bytes,
    record_batch_offset,
    record_batch_size,
  );
  return load_vectors(record_batch_bytes, schema);
};

/// `latest_batch` should have `id` (string), and `mem` (ArrayBuffer) fields.
Batch.prototype.sync = function (latest_batch, schema) {
  const markers = load_markers(latest_batch.mem);
  let header_offset = markers.header_offset;

  // extract the metaversion from the header of the latest batch
  // more informations on metaversions can be found on the `Metaversion` struct
  const dataview_latest = new DataView(latest_batch.mem);
  // note: the header consists of two 32-bit unsigned integers (aka `u32` in Rust)
  const latest_mem_version = dataview_latest.getUint32(header_offset, true);
  // we use header_offset + 4 here because the both integers have a width of 4
  // bytes
  const latest_batch_version = dataview_latest.getUint32(
    header_offset + 4,
    true,
  );

  const should_load = this.batch_version < latest_batch_version;

  if (this.mem_version < latest_mem_version) {
    if (!should_load) {
      throw new Error(
        "Should be impossible to have new memory without new batch",
      );
    }

    // JS is in same process as Rust, so just need to update pointer.
    this.mem = latest_batch.mem;
    this.mem_version = latest_mem_version;
  }

  if (should_load) {
    this.vectors = load_marked_vectors(this.mem, schema);
    this.cols = {}; // Reset columns because they might be invalid due to vectors changing.
    this.batch_version = latest_batch_version;
  }
};

/// `name` is the name of the Arrow column/field.
/// `loader` is optional.
Batch.prototype.load_col = function (name, loader) {
  const vector = this.vectors[name];
  if (!vector) throw new ReferenceError("Missing vector for " + name);

  let col;
  if (loader) {
    col = loader(vector);
  } else if (name.startsWith("_HIDDEN_") || name.startsWith("_PRIVATE_")) {
    // only agent-scoped fields are fully loaded by default
    // TODO: Whether a column is nullable doesn't currently affect Arrow loading in JS, but if we
    //       want to modify fixed-size non-nullable columns in place, `is_nullable` needs to be
    //       set to the correct value here.
    const is_nullable = undefined;
    col = hash_util.load_shallow(vector, is_nullable, vector.type.is_any);
  } else {
    // TODO: Whether a column is nullable doesn't currently affect Arrow loading in JS, but if we
    //       want to modify fixed-size non-nullable columns in place, `is_nullable` needs to be
    //       set to the correct value here.
    const is_nullable = undefined;
    col = hash_util.load_full(vector, is_nullable, vector.type.is_any);
  }
  return (this.cols[name] = col);
};

/// Load columns that are in `schema`, but haven't been loaded yet
/// (or were loaded, but then were erased again). Uses optional
/// custom loaders.
Batch.prototype.load_missing_cols = function (schema, loaders) {
  for (var i = 0; i < schema.fields.length; ++i) {
    const name = schema.fields[i].name;
    if (!this.cols[name]) this.load_col(name, loaders[name]);
  }
};

const array_data_from_col = (col, field_type) => {
  // TODO: Would `new arrow.Builder` work?
  // TODO: Faster way to convert JS array to vector than using `Builder`?
  //       `arrow.Vector.from(col)` doesn't seem to quite work.
  const builder = new arrow.makeBuilder({
    type: field_type,
    nullValues: [null, undefined],
  });

  // TODO: Is there a way to add the whole column all at once to `builder`?
  try {
    for (var i_agent = 0; i_agent < col.length; ++i_agent) {
      builder.append(col[i_agent]);
    }
  } catch (e) {
    throw new Error(
      "Flushing error: " +
        JSON.stringify([i_agent, col[i_agent], field_type, String(e)]),
    );
  }

  // JS Arrow doesn't really document what `builder.finish` does, but
  // maybe it affects later serialization.
  builder.finish();
  return builder.flush();
};

// TODO: Can JS Arrow silently coerce some flushed values to different types if
//       their type differs from what it's supposed to be according to the schema?
const ffi_data_from_array_data = (array_data) => {
  const child_data = [];
  for (var i = 0; i < array_data.children.length; ++i) {
    child_data[i] = ffi_data_from_array_data(array_data.children[i]);
  }

  const buffers = [];
  const offsets = array_data.valueOffsets;
  const values = array_data.values;
  if (offsets) {
    buffers[0] = offsets.buffer;
    if (values) buffers.push(values.buffer);
  } else if (values) {
    buffers[0] = values.buffer;
  }

  return {
    // Get datatype from schema later.
    len: array_data.length,
    null_count: array_data.nullCount,
    null_bits: array_data.nullBitmap.buffer,
    buffers: buffers,
    child_data: child_data,
  };
};

Batch.prototype.flush_changes = function (schema, skip) {
  const changes = [];
  // TODO: Benchmark vs `Object.entries` and vs `for (var col in cols)`.
  for (var i_field = 0; i_field < schema.fields.length; ++i_field) {
    const field = schema.fields[i_field];
    const col = this.cols[field.name];
    if (!col || skip[field.name] || col.length === 0) continue; // A package might not require all columns,
    // in which case some columns that are in the schema
    // might be missing from `cols`. (But columns that
    // are in `cols` should always be in schema too.)

    if (this.vectors[field.name].type.is_any) {
      for (var i_agent = 0; i_agent < col.length; ++i_agent) {
        col[i_agent] = JSON.stringify(col[i_agent]);
      }
    }
    const array_data = array_data_from_col(col, field.type);
    const data = ffi_data_from_array_data(array_data);
    changes.push({
      // Some fields might be skipped, so a field's index in `changes` might not be equal to `i_field`.
      i_field: i_field, //
      data: data,
    });
  }
  return changes;
};

export const Batches = function () {
  this.batches = {};
};

Batches.prototype.get = function (batch_id) {
  return this.batches[batch_id];
};

/// `latest_batch` should have `id` (string), `batch_version` (number),
/// `mem_version` (number) and `mem` (ArrayBuffer) fields.
Batches.prototype.sync = function (latest_batch, schema) {
  let loaded_batch = this.batches[latest_batch.id];
  if (!loaded_batch) {
    this.batches[latest_batch.id] = loaded_batch = new Batch();
  }
  loaded_batch.sync(latest_batch, schema);
  return loaded_batch;
};

File: packages/engine/lib/execution/src/runner/python/batch.py
import json
import struct

import pyarrow as pa

import hash_util
from wrappers import load_shared_mem
from wrappers import shared_buf_from_c_memory
from wrappers import dynamic_meta_from_c_memory
from wrappers import static_meta_from_schema
from wrappers import flush
from wrappers import unload_shared_mem

N_MARKER_BYTES = 8  # Markers are all u64s.
N_MARKERS = 8  # NUMBER_OF_MARKERS in datastore/memory.rs.
N_MARKERS_BYTES = N_MARKER_BYTES * N_MARKERS


def _load_markers_unchecked(mem):
    markers_bytes = mem[:N_MARKERS_BYTES].to_pybytes()

    # '<' implies little-endian.
    # '<' also implies standard sizes, so Q matches a u64.
    markers_fmt = "<" + "Q" * N_MARKERS
    return struct.unpack_from(markers_fmt, markers_bytes)


def load_markers(mem):
    markers = _load_markers_unchecked(mem)
    verify_markers(markers, mem)
    return markers


def verify_markers(markers, mem):
    assert len(markers) == N_MARKERS, markers

    # Same order as in `memory::shared_memory/markers.rs`.
    # Units are all numbers of bytes.
    (
        schema_offset,
        schema_size,
        header_offset,
        header_size,
        meta_offset,
        meta_size,
        data_offset,
        data_size,
    ) = markers

    # The "meta bytes" here do *not* contain the schema's key-value metadata.
    # They contain what is officially called a "RecordBatch message data header", but
    # the Rust implementation just calls it a "RecordBatch message".
    # https://arrow.apache.org/docs/format/Columnar.html#recordbatch-message

    # Schema comes immediately after markers.
    assert (
            schema_offset == N_MARKERS_BYTES
    ), f"schema_offset: {schema_offset}, n_markers_bytes: {N_MARKERS_BYTES}"
    assert schema_offset + schema_size <= header_offset
    assert header_offset + header_size <= meta_offset
    assert meta_offset + meta_size <= data_offset
    assert data_offset + data_size <= mem.size


def parse_any_type_fields(metadata):
    any_type_fields = set()

    # arrow2 serializes empty dictionaries as None (rather than `{}`, as arrow
    # did)
    if metadata is None:
        return any_type_fields

    field_names = metadata.get("any_type_fields")

    if field_names:
        for field_name in field_names.split(","):
            any_type_fields.add(field_name)

    return any_type_fields


def load_record_batch(mem, schema=None):
    (
        schema_offset,
        schema_size,
        _header_offset,
        _header_size,
        meta_offset,
        _meta_size,
        data_offset,
        data_size,
    ) = load_markers(mem)
    # Pyarrow exposes a function for parsing the record batch message data header and
    # record batch data together, but not functions for parsing them separately, so
    # they should be contiguous in memory. (Or have to use a hack to pretend that
    # what is between them is padding.)
    if schema is None:
        schema_buf = mem[schema_offset: schema_offset + schema_size]
        schema = pa.ipc.read_schema(schema_buf)

    record_batch_buf = mem[meta_offset: data_offset + data_size]
    record_batch = pa.ipc.read_record_batch(record_batch_buf, schema)

    any_type_fields = parse_any_type_fields(schema.metadata)
    return record_batch, any_type_fields


# Returns dataset name, dataset contents and whether JSON could be loaded.
def load_dataset(batch_id):
    mem = shared_buf_from_c_memory(load_shared_mem(batch_id))
    (_, _, header_offset, header_size, _, _, data_offset, data_size) = load_markers(mem)

    # The header has the shortname of the dataset
    n_metaversion_bytes = 8  # Memory u32 + batch u32 version
    name_offset = header_offset + n_metaversion_bytes  # Skip metaversion.
    header_end = header_offset + header_size
    name_buf = mem[name_offset:header_end]
    dataset_name = str(name_buf.to_pybytes().decode("utf-8"))

    # This data buffer has the dataset as a JSON string
    data_buf = mem[data_offset: data_offset + data_size]
    dataset_utf8 = data_buf.to_pybytes().decode("utf8")
    try:
        return dataset_name, json.loads(dataset_utf8), True
    except json.JSONDecodeError:
        # TODO: Extract parsing error line number from exception.
        return dataset_name, dataset_utf8, False


class Batch:
    def __init__(self, batch_id):
        self.id = batch_id

        self.mem_version = -1
        self.batch_version = -1
        # After loading, `mem` will be a shared buffer.
        self.mem = None
        # After loading, `record_batch` will be a record batch.
        self.record_batch = None
        # TODO: Remove `any_type_fields` after upgrading Arrow and putting metadata in individual
        #       columns.
        self.any_type_fields = None
        # Syncing erases columns that have become invalid.
        self.cols = {}

        # For flushing:
        self.c_memory = None
        self.dynamic_meta = None
        self.static_meta = None

    def load_persisted_metaversion(self):
        if self.mem is None:
            return 0, 0, None

        # Can't verify markers right now as `self.mem` maybe needs to be reloaded first (e.g. after
        # resizing)
        markers = _load_markers_unchecked(self.mem)
        (_, _, header_offset, _, _, _, _, _) = markers

        n_metaversion_bytes = 8  # Memory u32 + batch u32 version
        metaversion_buffer = self.mem[
                             header_offset: header_offset + n_metaversion_bytes
                             ]
        metaversion_bytes = metaversion_buffer.to_pybytes()
        mem_version, batch_version = struct.unpack_from("<II", metaversion_bytes)
        return mem_version, batch_version, markers

    def sync(self, batch, schema=None):
        (
            persisted_mem_version,
            persisted_batch_version,
            markers,
        ) = self.load_persisted_metaversion()

        should_load_batch = self.batch_version < persisted_batch_version
        if self.mem_version < persisted_mem_version:
            assert (
                should_load_batch
            ), "Should be impossible to have new memory without new batch"

            # `load_shared_mem` throws an exception if loading fails,
            # but otherwise the returned pointer to shared memory is non-null.
            self.c_memory = load_shared_mem(batch.id)
            self.mem = shared_buf_from_c_memory(self.c_memory)

            self.mem_version = persisted_mem_version

        if markers:
            verify_markers(markers, self.mem)

        if should_load_batch:
            self.record_batch, self.any_type_fields = load_record_batch(
                self.mem, schema
            )
            self.cols = {}  # Avoid using obsolete column data.
            self.static_meta = static_meta_from_schema(self.record_batch.schema)
            self.dynamic_meta = dynamic_meta_from_c_memory(self.c_memory)

            self.batch_version = persisted_batch_version

    def load_col(self, name, loader=None):
        i_field = self.record_batch.schema.get_field_index(name)
        if i_field < 0:
            raise RuntimeError(f"Missing field for {name}")

        field = self.record_batch.schema.field(i_field)
        vector = self.record_batch.column(i_field)
        is_any = name in self.any_type_fields
        if loader is not None:
            col = loader(vector, field.nullable, is_any)
        elif name.startswith("_PRIVATE_") or name.startswith("_HIDDEN_"):
            # only agent-scoped fields are fully loaded by default
            col = vector
        else:
            col = hash_util.load_full(vector, field.nullable, is_any)

        self.cols[name] = col
        return col

    # Load columns that are in `schema`, but haven't been loaded yet
    # (or were loaded, but then were erased again). Uses optional
    # custom loaders.
    def load_missing_cols(self, schema, loaders):
        for field_name in schema.names:
            if field_name not in self.cols:
                self.load_col(field_name, loaders.get(field_name))

    def flush_changes(self, schema, skip):
        any_type_fields = parse_any_type_fields(schema.metadata)

        # Dynamically accessed columns (if any) were added to `cols` by `state`.
        changes = []
        for field_name, col in self.cols.items():
            if field_name in skip or not isinstance(col, list) or len(col) == 0:
                # Assume that column wasn't written to or was writable in place.
                # TODO: More robust check for this (i.e. for shallow-loaded columns)
                continue

            i_field = schema.get_field_index(field_name)
            if i_field < 0:
                continue  # Not supposed to have this column in `cols`?

            field = schema.field(i_field)
            if field.name in any_type_fields:
                # Convert `any`-type array of JSON values to array of JSON strings
                # for Arrow serialization as a string column.
                py_col = [json.dumps(elem) for elem in col]
            elif isinstance(col[0], pa.Scalar):
                # Shallow-loaded column; can be modified in place
                continue
            else:
                # TODO: Custom loaders with intermediate level of shallow loading
                #       (These currently result in an exception from `pa.array`.)
                py_col = col

            changes.append(
                {"i_field": i_field, "data": pa.array(py_col, type=field.type)}
            )

        if len(changes) == 0:
            return

        self.batch_version += 1
        did_resize = flush(self.c_memory, self.dynamic_meta, self.static_meta, changes)
        if did_resize:
            # `c_memory` is updated inside `_flush_changes` if memory is resized.
            self.mem_version += 1
            self.mem = shared_buf_from_c_memory(self.c_memory)
            self.dynamic_meta = dynamic_meta_from_c_memory(self.c_memory)


class Batches:
    def __init__(self):
        self.batches = {}

    def get(self, batch_id):
        return self.batches[batch_id]

    def sync(self, latest_batch, schema=None):
        loaded_batch = self.batches.get(latest_batch.id)
        if loaded_batch is None:
            self.batches[latest_batch.id] = loaded_batch = Batch(latest_batch.id)

        # `loaded_batch` is changed in-place. Return is for convenience.
        loaded_batch.sync(latest_batch, schema)
        return loaded_batch

    def free(self):
        # TODO: Check that this releases references to shared memory
        #       (Call _free_rust_static_meta, _free_rust_dynamic_meta here?)
        #   see https://app.asana.com/0/1201461747883418/1201634225076144/f
        # TODO: Make this the `__del__` method?
        for batch in self.batches.values():
            unload_shared_mem(batch.c_memory)
        self.batches = {}
